src/AlsaPcm.cpp

```diff
@@ -543,4 +582,243 @@ snd_pcm_format_t AlsaPcm::convertPcmFormat(uint8_t format)
+
+{
+	snd_pcm_uframes_t minFrames = static_cast<snd_pcm_uframes_t>(req.buffer.min);
+	snd_pcm_uframes_t maxFrames = static_cast<snd_pcm_uframes_t>(req.buffer.max);
+	int ret;
+
+	if ((ret = snd_pcm_hw_params_set_buffer_size_minmax(mHwQueryHandle, mHwQueryParams,
+			&minFrames, &maxFrames)) < 0)
+	{
+		/*
+		 * This is not really a fatal error, the frontend just tries to
+		 * reduce configuration space.
+		 * But, anyway, throw now, so we can return error code to
+		 * the frontend.
+		 */
+	}
+
+	if ((ret = snd_pcm_hw_params_get_buffer_size_min(mHwQueryParams, &minFrames)) < 0)
+	{
+		throw Exception("Can't get buffer min" + mDeviceName, -ret);
+	}
+
+	if ((ret = snd_pcm_hw_params_get_buffer_size_max(mHwQueryParams, &maxFrames)) < 0)
+	{
+		throw Exception("Can't get buffer max" + mDeviceName, -ret);
+	}
+
+	resp.buffer.min = static_cast<unsigned int>(minFrames);
+	resp.buffer.max = static_cast<unsigned int>(maxFrames);
+}
+
+void AlsaPcm::queryHwParamChannels(SoundItf::PcmParamRanges& req, SoundItf::PcmParamRanges& resp)
+{
+	int ret;
+
+	if ((ret = snd_pcm_hw_params_set_channels_minmax(mHwQueryHandle, mHwQueryParams,
+			&req.channels.min, &req.channels.max)) < 0)
+	{
+		/*
+		 * This is not really a fatal error, the frontend just tries to
+		 * reduce configuration space.
+		 * But, anyway, throw now, so we can return error code to
+		 * the frontend.
+		 */
+	}
+
+	if ((ret = snd_pcm_hw_params_get_channels_min(mHwQueryParams, &resp.channels.min)) < 0)
+	{
+		throw Exception("Can't get channels min " + mDeviceName, -ret);
+	}
+
+	if ((ret = snd_pcm_hw_params_get_channels_max(mHwQueryParams, &resp.channels.max)) < 0)
+	{
+		throw Exception("Can't get channels max " + mDeviceName, -ret);
+	}
+}
+
+void AlsaPcm::queryHwParamPeriod(SoundItf::PcmParamRanges& req, SoundItf::PcmParamRanges& resp)
+{
+	snd_pcm_uframes_t minFrames = static_cast<snd_pcm_uframes_t>(req.period.min);
+	snd_pcm_uframes_t maxFrames = static_cast<snd_pcm_uframes_t>(req.period.max);
+	int ret;
+
+	if ((ret = snd_pcm_hw_params_set_period_size_minmax(mHwQueryHandle, mHwQueryParams,
+			&minFrames, 0, &maxFrames, 0)) < 0)
+	{
+		/*
+		 * This is not really a fatal error, the frontend just tries to
+		 * reduce configuration space.
+		 * But, anyway, throw now, so we can return error code to
+		 * the frontend.
+		 */
+	}
+
+	if ((ret = snd_pcm_hw_params_get_period_size_min(mHwQueryParams, &minFrames, 0)) < 0)
+	{
+		throw Exception("Can't get period min" + mDeviceName, -ret);
+	}
+
+	if ((ret = snd_pcm_hw_params_get_period_size_max(mHwQueryParams, &maxFrames, 0)) < 0)
+	{
+		throw Exception("Can't get period max" + mDeviceName, -ret);
+	}
+
+	resp.period.min = static_cast<unsigned int>(minFrames);
+	resp.period.max = static_cast<unsigned int>(maxFrames);
+}
+
+void AlsaPcm::queryHwParamFormats(SoundItf::PcmParamRanges& req, SoundItf::PcmParamRanges& resp)
+{
+	snd_pcm_format_mask_t* alsa_formats = nullptr;
+	int ret;
+
+	try {
+		alsa_formats = toAlsaFormatsMask(req.formats);
```

Comment:
Better to create and fill mask array right here with snd_pcm_format_mask_alloca. No need to free in two places.